nemo_aligner/models/nlp/gpt/gpt_steerlm_model.py

```diff
@@ -0,0 +1,269 @@
+        num_of_microbatches = gbs // mbs
+        group_batch = {k: v for k, v in batch.items() if isinstance(v, torch.Tensor)}
+
+        data_iter = get_iterator_k_split(group_batch, num_of_microbatches)
+
+        fwd_bwd_function = get_forward_backward_func()
+
+        # first get the forward nll loss for each of the microbatches
+        fwd_loss_fn = self.get_forward_output_only_for_weight_func()
+        losses_forward = fwd_bwd_function(
+            forward_step_func=fwd_loss_fn,
+            data_iterator=data_iter,
+            model=self.model,
+            num_microbatches=num_of_microbatches,
+            forward_only=True,
+            micro_batch_size=mbs,
+            seq_length=seq_length,
+        )
+        distance = None
+        if parallel_state.is_pipeline_last_stage():
+            updated_weight = torch.concatenate([i["loss"] for i in losses_forward])
+            base_weight = (
+                -updated_weight.reshape(-1, response_size) - batch["log(Q(y|a,x))"].reshape(-1, response_size).cuda()
+            )
+            base_weight_p = torch.nn.functional.softmax(base_weight, dim=-1)
+            ws_weight = batch["ws"].reshape(-1, response_size).cuda()
+            weight = ws_weight - base_weight_p
+            # group size is the number of responses generated
+            loss_mbs = int(self.cfg.steerlm2.micro_batch_size)
+            num_of_microbatches = gbs // loss_mbs
+            # split updated weights into groups of size response_size
+            num_of_batches = len(updated_weight) // response_size
+            avg_num_valid_tokens_in_ub = (
+                batch["loss_mask"]
+                .reshape(-1, response_size, seq_length)
+                .sum(axis=-1)
+                .float()
+                .mean(axis=-1, keepdim=True)
+                .repeat([1, response_size])
+            )
+            avg_num_valid_tokens_in_ub = avg_num_valid_tokens_in_ub / (num_of_microbatches / num_of_batches)
+
+            distance = ((ws_weight + 1e-8) * torch.log((ws_weight + 1e-8) / (base_weight_p + 1e-8))).sum(axis=1).mean()
+            distance = average_losses_across_data_parallel_group([distance])
+
+            batch["base_weight"] = base_weight_p.reshape(-1)
+            batch["weight"] = weight.reshape(-1)
+            batch["avg_num_valid_tokens_in_ub"] = avg_num_valid_tokens_in_ub.reshape(-1).cuda()
+        return distance, seq_length, gbs
+
+    def get_loss_and_metrics(self, batch, forward_only):
+        """Take a data_iter which is an interator over the microbatches
+            and return loss as well as metrics
+        """
+        distance, seq_length, gbs = self.compute_baselilne_weights(batch)
+
+        fwd_bwd_function = get_forward_backward_func()
+
+        set_sync_funcs(self, forward_only)
+        mbs = int(self.cfg.steerlm2.micro_batch_size)
+        assert gbs % mbs == 0, "Global batch size must be divisible by micro batch size"
+        number_of_microbatches = gbs // mbs
```

Comment:
`from megatron.core.utils import divide` will do what you want it checks the condition too

Reply:
changed